### PR TITLE
feat(dracut-install) Add fw_devlink suppliers as module dependencies

### PR DESCRIPTION
**Pull request by athierry1**
## Changes

Dracut currently finds kernel module dependencies using two methods: depmod and softdeps. However, these are often insufficient on embedded systems where a lot of driver dependencies are only described in the device tree. This is often the case for low-level devices such as clocks, regulators, pinctrls, etc.

This patch series allows dracut to find those device tree dependencies, by parsing the sysfs supplier nodes populated by fw_devlink.

This slows dracut down a little bit, however. Depending on whether dracut is called with `--hostonly` or not, I measured 25-35% slowdown on the Raspberry Pi 4. According to gperftools, most of the extra time is spent in `kmod_module_new_from_lookup()`

v2. Addressed review comments
v3.
- walk the sysfs once
- make sure that if one of the suppliers in a supplier chain is not a module (ie. is built-in), the suppliers it depends on are still processed

v4. Addressed review comments

## Checklist
- [x] I have tested it locally
- [ ] I have reviewed and updated any documentation if relevant
- [ ] I am providing new code and test(s) for it

### CODE REVIEW

**Comment by LaszloGombos** [comment]
CC @nabijaczleweli to help out with the review

**Comment by LaszloGombos** [comment]
@athierry1 Can you you help us and think of a why how this change can be tested ? Not asking to write the test case yourself, but maybe think of an environment or command line for testing ? Is this change makes a difference on x64 or only on arm ?

File: src/install/dracut-install.c [hunk 1]
```diff
@@ -1515,7 +1515,105 @@ static bool check_module_path(const char *path)
...
+                if (strcmp(ftsent->fts_name, "modalias") == 0) {
+                        char sysfs_node[PATH_MAX] = {0};
+                        memcpy(sysfs_node, ftsent->fts_parent->fts_path, ftsent->fts_parent->fts_pathlen);
+                        sysfs_node[ftsent->fts_parent->fts_pathlen] = '\0';
+
+                        _cleanup_kmod_module_unref_list_ struct kmod_list *list = NULL;
+                        struct kmod_list *l = NULL;
+
+                        if (find_module_from_sysfs_node(ctx, sysfs_node, &list) < 0)
+                                continue;
```
Comment (nabijaczleweli) [inline src/install/dracut-install.c:1591]: you keep taking FTS paths, copying them to slice them, then copying them *again* to unslice them.

you can avoid the clumsy copy here and below by getting rid of `sysfs_node` entirely and having
```c
if (find_module_from_sysfs_node(ctx, ftsent->fts_parent->fts_path, ftsent->fts_parent->fts_pathlen, &list) < 0)
```
coupled with
```c
static int find_module_from_sysfs_node(struct kmod_ctx *ctx, const char *sysfs_node, int sysfs_node_len, struct kmod_list **modules)
{
        char modalias_path[PATH_MAX] = {0};
        snprintf(modalias_path, sizeof(modalias_path), "%.*s/modalias", sysfs_node_len, sysfs_node);
        _cleanup_fclose_ FILE *modalias_file = NULL;
```
and similarly `find_suppliers_for_sys_node()` as
```c
find_suppliers_for_sys_node(ctx, suppliers, ftsent->fts_parent->fts_path, ftsent->fts_parent->fts_pathlen);
```
and
```c
find_suppliers_for_sys_node(ctx, suppliers, parent->fts_path, parent->fts_pathlen);
```
and have *that* do
```c
static void find_suppliers_for_sys_node(struct kmod_ctx *ctx, Hashmap *suppliers, const char *node_path_raw, size_t node_path_len)
{
        char node_path[PATH_MAX] = {};
        memcpy(node_path, node_path_raw, node_path_len);
        DIR *d = opendir(node_path);
        struct dirent *dir;
```

File: src/install/dracut-install.c [hunk 2]
```diff
@@ -1515,7 +1515,105 @@ static bool check_module_path(const char *path)
...
+                                if (strcmp(name, module) == 0) {
+                                        find_suppliers_for_sys_node(ctx, suppliers, sysfs_node);
+
+                                        // Also add suppliers of parent nodes
+                                        FTSENT *parent = ftsent->fts_parent;
+                                        while(parent->fts_level >= 1) {
+                                                memcpy(sysfs_node, parent->fts_path, parent->fts_pathlen);
+                                                sysfs_node[parent->fts_pathlen] = '\0';
+
+                                                find_suppliers_for_sys_node(ctx, suppliers, sysfs_node);
```
Comment (nabijaczleweli) [inline src/install/dracut-install.c:1606]: AFAICT this processes `ftsent->fts_parent` twice (once raw, aliased as sysfs_node, the other time in the first loop iteration)?

File: src/install/dracut-install.c [hunk 3]
```diff
@@ -1515,7 +1515,105 @@ static bool check_module_path(const char *path)
...
+                        kmod_list_foreach(l, list) {
+                                _cleanup_kmod_module_unref_ struct kmod_module *mod = kmod_module_get_module(l);
+                                const char *name = kmod_module_get_name(mod);
+
+                                if (strcmp(name, module) == 0) {
+                                        find_suppliers_for_sys_node(ctx, suppliers, sysfs_node);
+
+                                        // Also add suppliers of parent nodes
+                                        FTSENT *parent = ftsent->fts_parent;
+                                        while(parent->fts_level >= 1) {
```
Comment (nabijaczleweli) [inline src/install/dracut-install.c:1602]: this obfuscates the for loop which it is, probably better to write it as
```c
for(FTSENT *parent = ftsent->fts_parent; parent->fts_level >= 1; parent = parent->fts_parent)
    find_suppliers_for_sys_node(ctx, suppliers, parent->fts_path, parent->fts_pathlen);
```

File: src/install/dracut-install.c [hunk 4]
```diff
@@ -1515,7 +1515,105 @@ static bool check_module_path(const char *path)
...
+
+                        kmod_list_foreach(l, list) {
+                                _cleanup_kmod_module_unref_ struct kmod_module *mod = kmod_module_get_module(l);
+                                const char *name = kmod_module_get_name(mod);
+
+                                if (strcmp(name, module) == 0) {
+                                        find_suppliers_for_sys_node(ctx, suppliers, sysfs_node);
+
+                                        // Also add suppliers of parent nodes
+                                        FTSENT *parent = ftsent->fts_parent;
```
Comment (nabijaczleweli) [inline src/install/dracut-install.c:1601]: on my amd64 machine the outer-most loop yields:
```
/sys/devices/platform/PNP0C14:00/wmi_bus/wmi_bus-PNP0C14:00/0E7AF9F2-44A1-4C6F-A4B0-A7678480DA61
/sys/devices/platform/PNP0C14:00
/sys/devices/platform/reg-dummy
/sys/devices/platform/coretemp.1
/sys/devices/platform/dcdbas
/sys/devices/platform/platform-framebuffer.0
/sys/devices/platform/microcode
/sys/devices/platform/GHES.32994
/sys/devices/platform/dmi-ipmi-si.0
/sys/devices/platform/GHES.32992
/sys/devices/platform/GHES.49377
/sys/devices/platform/coretemp.0
/sys/devices/platform/ipmi_bmc.0
/sys/devices/platform/PNP0103:00
/sys/devices/platform/serial8250
/sys/devices/platform/i8042/serio0
/sys/devices/platform/i8042/serio1
/sys/devices/platform/i8042
/sys/devices/platform/GHES.32993
/sys/devices/platform/GHES.49378
/sys/devices/platform/GHES.227
/sys/devices/platform/pcspkr/input/input5
/sys/devices/platform/pcspkr
/sys/devices/platform/GHES.49376
```

Are you *sure* it's safe/sensible to process, for example, `/s/d/p/i8042` thrice? If not, mark `parent->fts_number = 1` on each iteration and make the loop condition `parent->level >= 1 && !parent->fts_number`.

File: src/install/dracut-install.c [hunk 5]
```diff
@@ -1515,7 +1515,105 @@ static bool check_module_path(const char *path)
...
+{
+        DIR *d;
+        struct dirent *dir;
+        d = opendir(node_path);
+        if (d) {
+                while ((dir = readdir(d)) != NULL) {
+                        if (strstr(dir->d_name, "supplier") != NULL) {
+                                char supplier_path[PATH_MAX] = {0};
+                                if (realpath(dir->d_name, supplier_path)) {
+                                        strncat(supplier_path, "/supplier", sizeof(supplier_path));
```
Comment (nabijaczleweli) [inline src/install/dracut-install.c:1555]: unchecked strncat

File: src/install/dracut-install.c [hunk 6]
```diff
@@ -1515,7 +1515,105 @@ static bool check_module_path(const char *path)
...
+
+                                        _cleanup_kmod_module_unref_list_ struct kmod_list *list = NULL;
+                                        struct kmod_list *l = NULL;
+
+                                        if (find_module_from_sysfs_node(ctx, supplier_path, &list) >= 0) {
+                                                kmod_list_foreach(l, list) {
+                                                        struct kmod_module *mod = kmod_module_get_module(l);
+                                                        char *supplier_module = strdup(kmod_module_get_name(mod));
+                                                        kmod_module_unref(mod);
+                                                        hashmap_put(suppliers, supplier_module, supplier_module);
```
Comment (nabijaczleweli) [inline src/install/dracut-install.c:1565]: this leaks `supplier_module` if `hashmap_put` errors

**Comment by athierry1** [comment]
@LaszloGombos the change should only make a difference on systems that use a device tree (so I guess only ARM systems, maybe also RISC-V?) 
So to test it, you could emulate an environment that's using a device tree. For instance a Raspberry Pi 3 emulated with QEMU (QEMU has a `raspi3b` machine)

You could :
- boot a standard raspbian image on this VM
- find a dependency in the device tree that's going to be your testcase (for example `cam1_regulator` (`fixed` module) references `firmware:expgpio` (module `gpio-raspberrypi-exp`) in the Raspberry Pi 3 device tree, so pulling the `fixed` module with `dracut-install` should also pull the `gpio-raspberrypi-exp` module)
- make sure those two drivers are built as modules (you might need to build a new kernel for that, I think most regulators, clks, etc. are built-in in the raspbian kernel)
- call `dracut-install --kerneldir=<above kernel dir> fixed` and check if `gpio-raspberrypi-exp` is also pulled in the output folder

I could try to make it work and write a little demo script if you're interested

**Comment by athierry1** [comment]
@nabijaczleweli thanks for your feedback, I'll try to submit a new version tomorrow

**Comment by LaszloGombos** [comment]
> I could try to make it work and write a little demo script if you're interested

Lets continue this on a separate issue - https://github.com/dracutdevs/dracut/issues/2082

**Comment by athierry1** [comment]
@nabijaczleweli I pushed a v2 addressing your feedback

File: src/install/dracut-install.c [hunk 7]
```diff
@@ -1515,7 +1515,114 @@ static bool check_module_path(const char *path)
...
+                                       struct kmod_list **modules)
+{
+        char modalias_path[PATH_MAX] = {0};
+        snprintf(modalias_path, sizeof(modalias_path), "%.*s/modalias", sysfs_node_len, sysfs_node);
+
+        _cleanup_close_ int modalias_file = -1;
+        if ((modalias_file = open(modalias_path, O_RDONLY | O_CLOEXEC)) == -1)
+                return 0;
+
+        char alias[getpagesize()];
```
Comment (nabijaczleweli) [inline src/install/dracut-install.c:1528]: any reason why not just `PAGE_SIZE`?

Comment (athierry1) [reply src/install/dracut-install.c:1528]: AFAICT `PAGE_SIZE` is only declared in the kernel headers. I came across https://stackoverflow.com/questions/37897645/page-size-undeclared-c that suggested to use `getpagesize()` instead 

But I see there's also a `page_size()` function defined in `src/install/util.c` that calls `sysconf(_SC_PAGESIZE)` Maybe I should use that instead, what do you think?

Comment (nabijaczleweli) [reply src/install/dracut-install.c:1528]: of course; POSIX requires `PAGESIZE`, glibc doesn't provide it. yeah, use that then.

File: src/install/dracut-install.c [hunk 8]
```diff
@@ -1515,7 +1515,114 @@ static bool check_module_path(const char *path)
...
+                                continue;
+
+                        kmod_list_foreach(l, list) {
+                                _cleanup_kmod_module_unref_ struct kmod_module *mod = kmod_module_get_module(l);
+                                const char *name = kmod_module_get_name(mod);
+
+                                if (strcmp(name, module) == 0) {
+                                        err = find_suppliers_for_sys_node(ctx, suppliers, sysfs_node->fts_path, sysfs_node->fts_pathlen);
+                                        if (err)
+                                                return err;
```
Comment (nabijaczleweli) [inline src/install/dracut-install.c:1605]: am I wrong to think this *always* overlaps with the first iteration of the loop below?

File: src/install/dracut-install.c [hunk 9]
```diff
@@ -1515,7 +1515,114 @@ static bool check_module_path(const char *path)
...
+{
+        char node_path[PATH_MAX] = {};
+        memcpy(node_path, node_path_raw, node_path_len);
+        DIR *d;
+        struct dirent *dir;
+        d = opendir(node_path);
+        if (d) {
+                while ((dir = readdir(d)) != NULL) {
+                        if (strstr(dir->d_name, "supplier") != NULL) {
+                                char supplier_path[PATH_MAX] = {0};
```
Comment (nabijaczleweli) [inline src/install/dracut-install.c:1546]: PATH_MAX is quite big to throw it around like this, you use the same buffer for node_path and supplier_path here.

File: src/install/dracut-install.c [hunk 10]
```diff
@@ -1515,7 +1515,114 @@ static bool check_module_path(const char *path)
...
+                                                kmod_list_foreach(l, list) {
+                                                        struct kmod_module *mod = kmod_module_get_module(l);
+                                                        char *supplier_module = strdup(kmod_module_get_name(mod));
+                                                        kmod_module_unref(mod);
+
+                                                        int err = hashmap_put(suppliers, supplier_module, supplier_module);
+                                                        if (err < 0) {
+                                                                free(supplier_module);
+                                                                return err;
+                                                        }
```
Comment (nabijaczleweli) [inline src/install/dracut-install.c:1568]: I don't think you need to `return err` here – hashmap_put() returns -EEXIST if `supplier_module` exists in the map already (the other error is ENOMEM but dracut-install doesn't care about allocations failing anyway) so just reduce to `if (hashmap_put()) free(supplier_module);` or equivalent and it's fine.

File: src/install/dracut-install.c [hunk 11]
```diff
@@ -1543,17 +1650,21 @@ static int install_dependent_module(struct kmod_module *mod, int *err)
                 _cleanup_kmod_module_unref_list_ struct kmod_list *modlist = NULL;
                 _cleanup_kmod_module_unref_list_ struct kmod_list *modpre = NULL;
                 _cleanup_kmod_module_unref_list_ struct kmod_list *modpost = NULL;
+                _cleanup_destroy_hashmap_ Hashmap *suppliers = hashmap_new(string_hash_func, string_compare_func);
                 log_debug("dracut_install '%s' '%s' OK", path, &path[kerneldirlen]);
                 install_firmware(mod);
-                modlist = kmod_module_get_dependencies(mod);
-                *err = install_dependent_modules(modlist);
+                *err = find_suppliers_for_module(ctx, suppliers, name);
```
Comment (nabijaczleweli) [inline src/install/dracut-install.c:1656]: if this really is *so much slower* (and that's a fair cop, you're walking a considerable hunk of sysfs for *every* module, AIUI), have you tried walking sysfs *once* and collecting all the module dependencies, then just consulting the resulting map here?

Comment (athierry1) [reply src/install/dracut-install.c:1656]: Thanks for the suggestion, I just tried it and it seems to speed things up considerably! I'll clean it up and send a new version that walks the sysfs only once.

I also found a bug in my code: if one of the suppliers in a supplier chain is not a module (ie. is built-in), the suppliers it depends on are not processed. So I'll fix that in the next version as well.

**Comment by athierry1** [comment]
Pushed a v3 that:
- walks the sysfs once
- makes sure that if one of the suppliers in a supplier chain is not a module (ie. is built-in), the suppliers it depends on are still processed

dracut-install is now a lot faster:
Executing `dracut-install -D ~/test -H --kerneldir /lib/modules/<my kernel> -o -m =drivers` on my RPi4 takes :
- 22s on the v2
- 1.26s on the v3

However, I didn't notice a speedup overall when calling dracut. It seems dracut calls dracut-install a lot of times, so the sysfs is still being walked a lot because the hashmap is recreated for each call.

**Comment by athierry1** [comment]
@nabijaczleweli could you please review the latest changes when you have the chance? Thank you!

File: src/install/dracut-install.c [hunk 12]
```diff
@@ -1515,7 +1516,110 @@ static bool check_module_path(const char *path)
...
+        DIR *d;
+        struct dirent *dir;
+        while (realpath(node_path, real_path) != NULL && strcmp(real_path, "/sys/devices")) {
+                d = opendir(node_path);
+                if (d) {
+                        while ((dir = readdir(d)) != NULL) {
+                                if (strstr(dir->d_name, "supplier:platform") != NULL) {
+                                        char supplier_path[PATH_MAX];
+                                        if (snprintf(supplier_path, sizeof(supplier_path), "%s/%s/supplier", real_path, dir->d_name) < sizeof(supplier_path)) {
+                                                char *real_supplier_path = realpath(supplier_path, NULL);
```
Comment (nabijaczleweli) [inline src/install/dracut-install.c:1575]: I've seen it when you posted it and the general approach appears sound in general. If it works it works.

You could just append to `real_path` here to not throw around a *third* PATH_MAX buffer, tho.

Comment (athierry1) [reply src/install/dracut-install.c:1575]: Thanks for reviewing!
I could be missing something, but `real_path` needs to be unchanged across iterations of the `while ((dir = readdir(d)) != NULL)` loop, so I think there is no way around having a third PATH_MAX buffer.

Comment (nabijaczleweli) [reply src/install/dracut-install.c:1575]: You could `size_t real_path_len = strlen(real_path)` after the `if(d)` and `snprintf(real_path + real_path_len, real_path - real_path_len, "/%s/supplier", dir->d_name)` I think?

Comment (athierry1) [reply src/install/dracut-install.c]: Oh this makes sense, thanks! I pushed a new version with this change.

**Comment by athierry1** [comment]
Hi @johannbg, did you get a chance to look at this PR?

**Comment by johannbg** [comment]
@athierry1 yup nothing stands out added it to the 059 milestone

**Comment by athierry1** [comment]
That's great, thank you!

**Comment by nabijaczleweli** [comment]
-2

**Comment by johannbg** [comment]
@athierry1 mark conversation resolved when they have been resolved dont edit your PR message with "$foo. Addressed review comments" 

For those of us that are extremely time constrained we ignore reviewing and merged PR that still have open conversation on them from other reviewers since we make the assumption that issues being pointed out by other reviewers have not been addressed

**Review by johannbg — APPROVED** [review]
LGTM

**Pull request merged**